### Navie Bays 

$$ P(A|B) = \frac {P(A).P(B|A)} {P(B)} $$


or 

$$ \text {Posterior Probability}  = \frac {\text(Prior Probability).\text(Likelihood)} {\text (Normalizing Constant)} $$
<p><b>Posterior Probability</b> - Which Represents the degree to which we believe a                       given model accurately describes the situation given the                       available data and all of our prior knowledge </p>

<p><b>Prior Probability</b> - Which Describe the Degree to which we believe the model
                   accurately describes reality based on all of our information <p>
    
<p> <b>Likelihood</b> - which describes how well the model predicts the data </p>
    
<p><b>Normalizing constant</b> - The constant that makes the posterior density                         INTEGRATE to one</p>
    

In [1]:
import pandas as pd
import numpy as np 

In [2]:
df  = pd.read_csv('navie_bays_play.csv')

outlook = "rainy", tempreature = "mild" , humidity="normal", windy = "True"


In [3]:
df

outlook temperature humidity  windy play
0      sunny         hot     high  False   no
1      sunny         hot     high   True   no
2   overcast         hot     high  False  yes
3      rainy        mild     high  False  yes
4      rainy        cool   normal  False  yes
5      rainy        cool   normal   True   no
6   overcast        cool   normal   True  yes
7      sunny        mild     high  False   no
8      sunny        cool   normal  False  yes
9      rainy        mild   normal  False  yes
10     sunny        mild   normal   True  yes
11  overcast        mild     high   True  yes
12  overcast         hot   normal  False  yes
13     rainy        mild     high   True   no

$$ P({A} |{B}) = \frac {P(A)P(B|A)} {P(B)}$$

$$ P(A) = \text{P(play=yes) } $$
$$ P(B)  = P(outlook=rainy)P(tempreature=mild)P(humidity=normal)P(Wind=strong) $$

In [4]:
play_yes = df['play'][df[ 'play'] == 'yes'].count()
play_total = df['play'].count()

In [5]:
play_yes

9

In [6]:
play_total

14

In [7]:
p_play_yes = play_yes / play_total 

In [8]:
p_play_yes

0.6428571428571429

In [9]:
p_play_no = 1 - p_play_yes
p_play_no

0.3571428571428571

In [10]:
outlook_rainy = df['outlook'][df['outlook'] =='rainy'].count()
outlook_total = df['outlook'].count()
outlook_rainy


5

In [12]:
p_outlook_rainy = outlook_rainy /outlook_total

In [13]:
p_outlook_rainy

0.35714285714285715

In [14]:
p_outlook_no_rainy = 1 - p_outlook_rainy

In [15]:
p_outlook_no_rainy

0.6428571428571428

In [16]:
tempreature_mild = df['temperature'][df['temperature']=='mild'].count()
tempreature_mild

6

In [17]:
total = df['outlook'].count()
total

14

In [18]:
p_tempreature_mild = tempreature_mild / total

In [19]:
p_tempreature_mild

0.42857142857142855

In [20]:
p_tempreature_no_mild = 1- p_tempreature_mild

In [21]:
p_tempreature_no_mild

0.5714285714285714

In [22]:
humidty_normal = df['humidity'][df['humidity']=='normal'].count() 
humidty_normal

7

In [23]:
p_humidty_normal = humidty_normal / total

In [24]:
p_humidty_normal

0.5

In [25]:
p_humidty_no_normal= 1  - p_humidty_normal

In [26]:
windy_true = df['windy'][df['windy']==True].count()
windy_true

6

In [27]:
p_windy_true = windy_true / total
p_windy_true

0.42857142857142855

In [28]:
p_windy_not_true = 1 - p_windy_true

In [29]:
p_windy_not_true

0.5714285714285714

In [30]:
 outlook_rainy_yes = df['outlook'][(df["outlook"]=='rainy') & (df["play"]=='yes') ].count()

In [31]:
p_outlook_rainy_yes =outlook_rainy_yes / df['play'][df['play'] == "yes"].count()
p_outlook_rainy_yes

0.3333333333333333

In [32]:
df['outlook'] = df['outlook'].map({
    'sunny':1, 'rainy':2, 'overcast':3
})

In [33]:
df['temperature'] = df['temperature'].map({
    'hot':1, 'mild':2,'cool':3
})

In [34]:
df['humidity'] = df['humidity'].map({
    'high':1,'normal':2
})

In [35]:
df['windy'] = df['windy'].map({
    True:1,False:2
})

In [36]:
df

outlook  temperature  humidity  windy play
0         1            1         1      2   no
1         1            1         1      1   no
2         3            1         1      2  yes
3         2            2         1      2  yes
4         2            3         2      2  yes
5         2            3         2      1   no
6         3            3         2      1  yes
7         1            2         1      2   no
8         1            3         2      2  yes
9         2            2         2      2  yes
10        1            2         2      1  yes
11        3            2         1      1  yes
12        3            1         2      2  yes
13        2            2         1      1   no

In [83]:
from sklearn.naive_bayes import MultinomialNB

In [84]:
clf = MultinomialNB()



In [85]:
clf.fit(df.drop('play',axis=1),df['play'])

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [87]:
df.columns

Index(['outlook', 'temperature', 'humidity', 'windy', 'play'], dtype='object')

In [88]:
data = [
    
    [2,2,2,1],
    [1,2,1,2]
]

In [89]:
clf.predict(data)

array(['yes', 'yes'], dtype='<U3')

#### Probabillity that we can play the game.

$$ P(Play=yes) = ? $$
$$ P(Play=no) = ? $$


In [14]:
c_yes = df.groupby(by='play')['outlook'].count()

In [15]:
c_yes

play
no     5
yes    9
Name: outlook, dtype: int64

$$ \text{P(play=yes)} =  {\frac {9} {14} }  = 0.0642$$ 
$$ \text{P(play=no)} = {\frac {5} {14} } = 0.3571$$ 

###### Probability We can play the game given ?
    Outlook = Sunny
    Temp = Cool
    Humidity = Hig
    Wind = Strong
    Play = yes

    P(Outlook=Sunny|Play=Yes) = 2 / 9
    P(Temp=Cool|Play=Yes) = 3/9
    P(Humidity=High|Play=yes) = 3/9
    P(Wind=Strong|Play=yes) =  3/9
    P(Play=Yes) = 9/14
    

###### Probability We can not play the game ? 

    P(Outlook=Sunny|Play=No) = 3 / 5
    P(Temp=Cool|Play=No) = 1/5
    P(Humidity=High|Play=No) = 4/5
    P(Wind=Strong|Play=No) =  3/5
    P(Play=No) = 5/14

P(X|C)P(C) or 

    P(X|Play=Yes)P(Play=Yes) = (2/9)*(3/9)*(3/9)*(3/9)*(9/14) = 0.0053
    P(X|Play=No)P(Play=No) =   (3/5)*(1/5)*(4/5)*(3/5)*(5/14) = 0.0206
    
    
    P(X) = P(Outlook=Sunny)*P(Temperature=Cool)*P(Humidity=High)*P(Wind=Strong)
    P(X) = (5/14)*(4/14)*(7/14)*(6/14)
    P(X) = 0.02186
    
    Dividing the results by this value : 
    
    P(Play=yes|X) = 0.0053/0.02186 = 0.2424
    P(Play=No|X) = 0.0206/0.02186 = 0.9421

In [58]:
clf = NavieBays()

In [59]:
clf.train(df.iloc[:,:-1],df.iloc[:,-1])

yes    9
no     5
Name: play, dtype: int64


In [56]:
a = np.random.randn(20,5)

In [36]:
clf.train(a[:,:-1],a[:,-1])

RangeIndex(start=0, stop=4, step=1)


In [60]:
x  = df['humidity'].value_counts()

In [65]:
f = list(zip(x.index,x))

In [66]:
f

[('normal', 7), ('high', 7)]

# text --> numerical 

In [90]:
msg1 = "hello i am good boy" # ham 
msg2 = "free car click now" # spam

In [91]:
from sklearn.feature_extraction.text import CountVectorizer

In [92]:
obj = CountVectorizer()

In [93]:
obj

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)

In [94]:
data = obj.fit_transform([msg1,msg2])

In [95]:
data

<2x8 sparse matrix of type '<class 'numpy.int64'>'
	with 8 stored elements in Compressed Sparse Row format>

In [96]:
data.data

array([1, 1, 1, 1, 1, 1, 1, 1], dtype=int64)

In [98]:
obj.vocabulary_ # Bag of words 

{'hello': 6,
 'am': 0,
 'good': 5,
 'boy': 1,
 'free': 4,
 'car': 2,
 'click': 3,
 'now': 7}

In [101]:
print(obj.fit_transform(["hello world how are you ","hello hello hello bye",
                        "let's hello to bye bye to hello good to go "]))

  (0, 9)	1
  (0, 0)	1
  (0, 5)	1
  (0, 8)	1
  (0, 4)	1
  (1, 1)	1
  (1, 4)	3
  (2, 2)	1
  (2, 3)	1
  (2, 7)	3
  (2, 6)	1
  (2, 1)	2
  (2, 4)	2


In [102]:
obj.vocabulary_

{'hello': 4,
 'world': 8,
 'how': 5,
 'are': 0,
 'you': 9,
 'bye': 1,
 'let': 6,
 'to': 7,
 'good': 3,
 'go': 2}

### IRIS --> MultinomialNB

In [103]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

In [104]:
from sklearn import datasets

In [105]:
iris = datasets.load_iris()

In [106]:
features = iris['data']
target = iris.target

In [108]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(features,target)

In [109]:
clf = MultinomialNB()

In [110]:
clf.fit(X_train,y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [111]:
pred = clf.predict(X_test)

In [113]:
accuracy_score(y_test,pred)

0.8421052631578947

In [114]:
confusion_matrix(y_test,pred)

array([[13,  0,  0],
       [ 0, 11,  0],
       [ 0,  6,  8]], dtype=int64)